torch/_dynamo/variables/misc.py

```diff
@@ -821,15 +823,23 @@ def call_function(
             unimplemented(
                 f"numpy dtype function is not supported yet. Got type {type(self.value)}."
             )
-        elif hasattr(torch_np, self.value.__name__):
+        elif any(hasattr(module, self.value.__name__) for module in SUPPORTED_MODULES):
             # TODO(larryliu0820): currently assuming all numpy.* functions are returning a ndarray that can be
             #  wrapped by NumpyNdarrayVariable which is wrong!
             getattr_fn = get_custom_getattr(torch_np)
             if getattr_fn:
                 unimplemented(
                     "torch_np has custom getattr implementation and dynamo doesn't support it"
                 )
-            func = getattr(torch_np, self.value.__name__)
+
+            for module in SUPPORTED_MODULES:
```

Comment:
Is there a better way to do this?

Reply:
In addition to that, can we extract this logic into a separate method? It's becoming more complicated. Would be good to have this util that takes in `self.value` and emits `func`, or throws proper exception.

Reply:
Also cc @voznesenskym

Reply:
Iterating here seems a bit wacky to me. Why is this not just a straight dict lookup?

Reply:
On a second thought, the module should be somewhere. Otherwise we couldn't tell apart, say, `torch.norm` and `torch.linalg.norm`!

Reply:
At this point, `self.value` is a numpy function and numpy doesn't seem to consistently have `__module__` on all the functions (seems like `ufuncs` don't have `__module__`) set. (Probably a numpy bug)

```python
>>> import numpy as np
>>> np.sin.__module__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'numpy.ufunc' object has no attribute '__module__'
>>> np.mean
<function mean at 0x7f067bf571f0>
>>> np.mean.__module__
'numpy'
```

So, I don't think we can rely on it.

Reply:
what about trying to get the module, and:
- If it has it, then fetch it from the correct module
- If it doesn't have it, then fetch the function from the main `np` module
?

I reckon that functions that live in a NumPy module do have their `__module__` attribute correctly set.

Reply:
Just found that `np.random.randn.__module__` returns `None`. I don't think information from `__module__` is reliable.

So, we need something like
```python
def get_torch_np_func(fn):
    import torch_np

    try:
        module = fn.__module__
        if not module:
            # Not sure if other submodules also do this.
            module = 'numpy.random'
    except AttributeError:
        module = 'numpy'

    module = module.replace('numpy', 'torch_np')
    module_qualified_name = module + "." + fn.__name__
    obj = torch_np

    try:
        # first is always `torch_np`, so we skip it.
        for path in module_qualified_name.split(".")[1:]:
            obj = getattr(obj, path)
        return obj
    except AttributeError:
        unimplemented(f"Can't find numpy function {fn} in torch_np")

```

Reply:
This HORRIBLE implementation is able to get the module of all the numpy functions we implement (and more):
```python
def get_np_module(fn):
    if isinstance(fn, ufunc):
        return "numpy"
    module = getattr(fn, "__module__", None)
    if module is None:
        s = str(fn)
        s = s.split(" ")
        # Handles objects like
        # <built-in method randn of numpy.random.mtrand.RandomState object at 0x7ff67319a940>
        if len(s) == 8:
            s = s[4]
            s = s.split(".")
            if s[0] == "numpy":
                module = f"{s[0]}.{s[1]}"
    elif "numpy.random" in module:
        # Handles np.random.MT19937.__module__ == 'numpy.random._mt19937.MT19937'
        # There's many of this form in torch.randn
        module = "numpy.random"
    return module
```

The test:
```python
def get_fns(mod):
    for fn in dir(mod):
        if fn[0] != "_":
            f = getattr(mod, fn)
            if callable(f):
                yield f


def test_module(mod):
    for fn in get_fns(mod):
        if (module := get_np_module(fn)) != mod.__name__:
            print(fn, module)



import numpy as np
test_module(np)
import numpy.fft as fft
test_module(fft)
import numpy.linalg as linalg
test_module(linalg)
import numpy.random as random
test_module(random)
```

Reply:
I was planning to create this map once and use it later for indexing.
```python
    from ..utils import NP_SUPPORTED_MODULES, NP_TO_TORCH_NP_MODULE

    # NP_SUPPORTED_MODULES = (np, np.fft, np.linalg, np.random)
    # NP_TO_TORCH_NP_MAP = {
    #    np: torch_np,
    #    np.fft: torch_np.fft,
    #    np.linalg: torch_np.linalg,
    #    np.random: torch_np.random,
    # }

    def get_np_to_torch_np_map():
        np_fn_to_torch_np_fn = {}

        for mod in NP_SUPPORTED_MODULES:
            np_fn_to_torch_np_fn.update(
                {
                    # `torch_np` doesn't implement functions like
                    # like `numpy.fromstring`, `numpy.fromfile`, etc.
                    v: getattr(NP_TO_TORCH_NP_MODULE[mod], k, None)
                    for k, v in mod.__dict__.items()
                    if callable(v)
                    and (getattr(v, "__module__", None) or mod.__name__) == mod.__name__
                }
            )

        return np_fn_to_torch_np_fn

    NP_FN_TO_TORCH_NP_FN = get_np_to_torch_np_map()
```

Reply:
that's certainly better, yep.